subprojects/groovy-json/src/main/java/groovy/json/JsonOutput.java

```diff
@@ -601,4 +280,966 @@ public String toString() {
+            CharBuf buffer = CharBuf.create(255);
+            writeObject(object, buffer); // checking null inside
+
+            return buffer.toString();
+        }
+
+        /**
+         * @see JsonOutput#toJson(Map)
+         */
+        public String toJson(Map m) {
+            if (m == null) {
+                return nullValue;
+            }
+
+            if (shouldExcludeType(m.getClass())) {
+                return EMPTY_VALUE;
+            }
+
+            CharBuf buffer = CharBuf.create(255);
+            writeMap(m, buffer);
+
+            return buffer.toString();
+        }
+
+        /**
+         * Serializes Number value and writes it into specified buffer.
+         */
+        private void writeNumber(Class<?> numberClass, Number value, CharBuf buffer) {
+            if (numberClass == Integer.class) {
+                buffer.addInt((Integer) value);
+            } else if (numberClass == Long.class) {
+                buffer.addLong((Long) value);
+            } else if (numberClass == BigInteger.class) {
+                buffer.addBigInteger((BigInteger) value);
+            } else if (numberClass == BigDecimal.class) {
+                buffer.addBigDecimal((BigDecimal) value);
+            } else if (numberClass == Double.class) {
+                Double doubleValue = (Double) value;
+                if (doubleValue.isInfinite()) {
+                    throw new JsonException("Number " + value + " can't be serialized as JSON: infinite are not allowed in JSON.");
+                }
+                if (doubleValue.isNaN()) {
+                    throw new JsonException("Number " + value + " can't be serialized as JSON: NaN are not allowed in JSON.");
+                }
+
+                buffer.addDouble(doubleValue);
+            } else if (numberClass == Float.class) {
+                Float floatValue = (Float) value;
+                if (floatValue.isInfinite()) {
+                    throw new JsonException("Number " + value + " can't be serialized as JSON: infinite are not allowed in JSON.");
+                }
+                if (floatValue.isNaN()) {
+                    throw new JsonException("Number " + value + " can't be serialized as JSON: NaN are not allowed in JSON.");
+                }
+
+                buffer.addFloat(floatValue);
+            } else if (numberClass == Byte.class) {
+                buffer.addByte((Byte) value);
+            } else if (numberClass == Short.class) {
+                buffer.addShort((Short) value);
+            } else { // Handle other Number implementations
+                buffer.addString(value.toString());
+            }
+        }
+
+        private void writeObject(Object object, CharBuf buffer) {
+            writeObject(null, object, buffer);
+        }
+
+        /**
+         * Serializes object and writes it into specified buffer.
+         */
+        private void writeObject(String key, Object object, CharBuf buffer) {
```

Comment:
Also if we do make the constructor `protected` making it possible to overload individual `write*` methods make sense?